### PR TITLE
Feature/return url message handling

### PR DESCRIPTION
**Pull request by ekkinox**
* Added possibility to customise tools launch error handling

Target release: `3.4.0`

### CODE REVIEW

File: Security/Exception/LtiToolMessageExceptionHandler.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```
Comment (peetya) [inline Security/Exception/LtiToolMessageExceptionHandler.php]: Now it's 2021 ;)

File: Security/Exception/LtiToolMessageExceptionHandlerInterface.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace OAT\Bundle\Lti1p3Bundle\Security\Exception;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Throwable;
+
+interface LtiToolMessageExceptionHandlerInterface
+{
+    public function handle(Throwable $exception, Request $request): Response;
```
Comment (peetya) [inline Security/Exception/LtiToolMessageExceptionHandlerInterface.php:31]: In LTI 1.3 context, the Response should be always `RedirectResponse`, no?

File: Security/Exception/LtiToolMessageExceptionHandler.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+use Lcobucci\JWT\Parser;
+use OAT\Library\Lti1p3Core\Message\Payload\LtiMessagePayload;
+use OAT\Library\Lti1p3Core\Security\Jwt\AssociativeDecoder;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Throwable;
+
+class LtiToolMessageExceptionHandler implements LtiToolMessageExceptionHandlerInterface
```
Comment (peetya) [inline Security/Exception/LtiToolMessageExceptionHandler.php:33]: It would be great to cover this class with tests. WDYT?

**Review by peetya — CHANGES_REQUESTED** [review]

File: Security/Exception/LtiToolMessageExceptionHandlerInterface.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
```
Comment (m-lagarde) [inline Security/Exception/LtiToolMessageExceptionHandlerInterface.php]: 2021 :smiling_imp:

File: Security/Exception/LtiToolMessageExceptionHandlerInterface.php (hunk 2 above)
Comment (ekkinox) [reply Security/Exception/LtiToolMessageExceptionHandlerInterface.php:31]: You can decide to interrupt the flow, and handle yourself error displaying.

So typhint with Response allow us regular and redirect responses, more flexible :)

File: Security/Exception/LtiToolMessageExceptionHandler.php (hunk 3 above)
Comment (ekkinox) [reply Security/Exception/LtiToolMessageExceptionHandler.php:33]: it is already, see coverage (still 100%) here: https://coveralls.io/builds/36829739 
it's covered in integration / functional tests

**Review by peetya — APPROVED** [review]

File: Security/Exception/LtiToolMessageExceptionHandler.php (hunk 1 above)
Comment (ekkinox) [reply Security/Exception/LtiToolMessageExceptionHandler.php]: indeed, fixed.

File: Security/Exception/LtiToolMessageExceptionHandlerInterface.php (hunk 4 above)
Comment (ekkinox) [reply Security/Exception/LtiToolMessageExceptionHandlerInterface.php]: indeed, fixed

**Review by m-lagarde — APPROVED** [review]

**Pull request merged**